File: backend/app/schemas/prescriptions.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.models.clinical import PrescriptionStatus


class DrugResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    atc_code: Optional[str] = None
    inn: str
    trade_name: Optional[str] = None
    form: Optional[str] = None
    dosage: Optional[str] = None


class PrescriptionCreate(BaseModel):
    encounter_id: uuid.UUID
    drug_id: uuid.UUID
    dosage: Optional[str] = None
    frequency: Optional[str] = None
    duration_days: Optional[int] = None
    quantity: Optional[int] = None
    instructions: Optional[str] = None


class PrescriptionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    doctor_id: uuid.UUID
    drug_id: uuid.UUID
    dosage: Optional[str] = None
    frequency: Optional[str] = None
    duration_days: Optional[int] = None
    quantity: Optional[int] = None
    instructions: Optional[str] = None
    status: PrescriptionStatus
    esoz_request_id: Optional[uuid.UUID] = None
    esoz_request_number: Optional[str] = None
    expires_at: Optional[datetime] = None
    created_at: datetime
    drug: Optional[DrugResponse] = None


class PrescriptionCancelRequest(BaseModel):
    reason: str

File: frontend/src/api/prescriptions.ts
import client from './client'
import type { PrescriptionResponse, DrugResponse, PrescriptionStatus } from './types'

export const prescriptionsApi = {
  create: (data: {
    encounter_id: string
    drug_id: string
    dosage?: string
    frequency?: string
    duration_days?: number
    quantity?: number
    instructions?: string
  }) => client.post<PrescriptionResponse>('/prescriptions', data),

  searchDrugs: (q = '', limit = 10) =>
    client.get<DrugResponse[]>('/prescriptions/drugs/search', { params: { q, limit } }),

  getByPatient: (patientId: string, status?: PrescriptionStatus) =>
    client.get<PrescriptionResponse[]>(`/prescriptions/patients/${patientId}/prescriptions`, {
      params: status ? { status } : undefined,
    }),

  get: (id: string) => client.get<PrescriptionResponse>(`/prescriptions/${id}`),

  cancel: (id: string, reason: string) =>
    client.patch<PrescriptionResponse>(`/prescriptions/${id}/cancel`, { reason }),
}
